include/deal.II/matrix_free/mapping_info.templates.h

```diff
@@ -670,7 +670,19 @@ namespace internal
           {
             typename dealii::Triangulation<dim>::cell_iterator cell_it(
               &tria, cells[j].first, cells[j].second);
-            fe_val.reinit(cell_it);
+            hp_fe_val.reinit(cell_it,
+                             numbers::invalid_unsigned_int,
+                             numbers::invalid_unsigned_int,
+                             active_fe_index);
```

Comment:
Unfortunately, we only have access here to the triangulation and not the dofhandler, so we need write it manually here.

Reply:
But there is no `DoFHandler` and `FiniteElement` in this evaluation at all, so to me this looks like a bug of `hp::FEValues` when initialized with a tria cell iterator and it should be fixed there. That said, you could probably just pass in 0 for the active fe index because there really should not be any.

Reply:
To be more specific, shouldn't we use the active fe index for the quadrature formula and mapping object, rather than the "fe" index, see also the argument list here: https://github.com/dealii/dealii/blob/e4ad46dc23fecadc80f4e015b18db2ab8e2fbb8c/include/deal.II/hp/fe_values.h#L367-L378

Reply:
I must admit I would prefer to have the same decision logic in the `reinit()` function that takes a iterator form a triangulation and that takes a iterator from a dofhandler. 

```cpp
  template <int dim, int spacedim>
  void
  FEValues<dim, spacedim>::reinit(
    const typename Triangulation<dim, spacedim>::cell_iterator &cell,
    const unsigned int                                          q_index,
    const unsigned int                                          mapping_index,
    const unsigned int                                          fe_index)
  {
    // determine which indices we
    // should actually use
    unsigned int real_q_index       = q_index;
    unsigned int real_mapping_index = mapping_index;
    unsigned int real_fe_index      = fe_index;


    if (real_q_index == numbers::invalid_unsigned_int)
      {
        if (real_fe_index == numbers::invalid_unsigned_int ||
            this->q_collection.size() == 1)
          real_q_index = 0;
        else
          real_q_index = fe_index;
      }


    if (real_mapping_index == numbers::invalid_unsigned_int)
      {
        if (real_fe_index == numbers::invalid_unsigned_int ||
            this->mapping_collection->size() == 1)
          real_mapping_index = 0;
        else
          real_mapping_index = fe_index;
      }


    if (real_fe_index == numbers::invalid_unsigned_int)
      real_fe_index = 0;


    // some checks
    AssertIndexRange(real_q_index, this->q_collection.size());
    AssertIndexRange(real_mapping_index, this->mapping_collection->size());
    AssertIndexRange(real_fe_index, this->fe_collection->size());


    // now finally actually get the
    // corresponding object and
    // initialize it
    this->select_fe_values(real_fe_index, real_mapping_index, real_q_index)
      .reinit(cell);
  }
```
Of course we could have this logic also in `MatrixFree` (as it is currently the case for cells for the quadrature rules, where we select either quadrature index `0` or `active_fe_index`). But since we now also add `MappingQCollection` and will need the same logic for faces I would suggest to handle this at a central place.

Reply:
I am not generally opposed to having this logic, but to me it completely goes against intuition to make automatic choices about some indices of an `FEValues` object that is not associated with a `DoFHandler` and a genuine `FiniteElement`. At the very least, we would need some careful comments in the `reinit` function and at the place of usage in `MatrixFree` because nobody will remember what happens without looking at the source code.

In my opinion, it would be much better to slightly break symmetry with the version taking `DoFHandler:cell_iterators` to make deductions for say the quadrature index instead, because that is an index that actually makes sense. But I admit that even that choice has a chance of getting misunderstood, so this would need comments as well.

Either way, I would be happy to get convinced of another way.